### PR TITLE
DOC: Add real docs.

### PR DESCRIPTION
**Pull request by ssanderson**

### CODE REVIEW

File: codetransformer/instructions.py [hunk 1]
```diff
@@ -11,7 +17,7 @@
 from .utils.no_default import no_default
 
 
-__all__ = ['Instruction'] + list(opmap)
+__all__ = ['Instruction'] + sorted(list(opmap))
```
Comment (llllllllll) [inline codetransformer/instructions.py:14]: we don't need to list this if we are caling sorted on it

File: codetransformer/instructions.py [hunk 2]
```diff
@@ -64,7 +70,7 @@ def _vartype(self):
 
 
 class InstructionMeta(ABCMeta, matchable):
-    _marker = object()  # sentinel
+    _marker = type('_marker', (object,), {})  # sentinel
```
Comment (llllllllll) [inline codetransformer/instructions.py]: why did you change this?

File: codetransformer/instructions.py [hunk 3]
```diff
@@ -207,10 +213,32 @@ def from_bytes(cls, bs):
...
+        """
+        The net effect of executing this instruction on the interpreter stack.
+
+        Instructions that pop values off the stack have negative stack effect
+        equal to the number of popped values.
+
+        Instructions that push values onto the stack have positive stack effect
+        equal to the number of popped values.
+
+        Example
```
Comment (llllllllll) [inline codetransformer/instructions.py]: missing the example.

File: codetransformer/transformers/__init__.py [hunk 4]
```diff
@@ -5,5 +11,4 @@
 __all__ = [
     'asconstants',
     'literals',
-    'optimize',
```
Comment (llllllllll) [inline codetransformer/transformers/__init__.py:8]: thats old, wow

File: codetransformer/transformers/constants.py [hunk 5]
```diff
@@ -60,14 +70,14 @@ class asconstants(CodeTransformer):
     >>> f()
     1
     """
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *builtin_names, **kwargs):
```
Comment (llllllllll) [inline codetransformer/transformers/constants.py:98]: thanks

File: codetransformer/transformers/constants.py [hunk 6]
```diff
@@ -49,6 +49,16 @@ class asconstants(CodeTransformer):
     """
     A code transformer that inlines names as constants.
```
Comment (llllllllll) [inline codetransformer/transformers/constants.py:50]: should we mention that if there are no arguments it inlines the whole builtins namespace?

File: codetransformer/transformers/literals.py [hunk 7]
```diff
@@ -47,17 +47,17 @@ class overloaded_dicts(CodeTransformer):
     >>> f()
     OrderedDict([('a', 1), ('b', 2), ('c', 3)])
     """
-    def __init__(self, astype):
+    def __init__(self, tnfm):
```
Comment (llllllllll) [inline codetransformer/transformers/literals.py]: I am not sure that this name is better

File: codetransformer/transformers/literals.py [hunk 8]
```diff
@@ -216,23 +222,60 @@ def overloaded_constants(type_):
...
+            tnfm("some string")
+
+        Parameters
+        ----------
+        tnfm : callable
+            Function to call on all string literals in the transformer target.
+
+        Examples
+        --------
+        >>> @overloaded_strs(lambda x: "ayy lmao ")
```
Comment (llllllllll) [inline codetransformer/transformers/literals.py:263]: ayy

**Comment by llllllllll** [comment]
These look so good! It's basically a real project now!

File: codetransformer/instructions.py (hunk 2 above)
Comment (ssanderson) [inline codetransformer/instructions.py]: I thought it might be causing the issues we were seeing with Instruction not being discovered correctly by sphinx, but that wasn't the case.

I left it because it seems slightly odd to have a random non-type object in Instruction's mro.

Comment (llllllllll) [inline codetransformer/instructions.py]: It never makes it into the mro, if it is the sentinel it is replaced with object. I explicitly made this a non type to avoid this confusion. You cannot be a subclass of a non-type object.

Comment (llllllllll) [inline codetransformer/instructions.py]: I did find a bug in the repr of the `Instruction` class, I will open a pr

File: codetransformer/transformers/literals.py (hunk 7 above)
Comment (ssanderson) [inline codetransformer/transformers/literals.py]: IMO, `astype` implies that the callable should be a type. `tnfm` I think is slightly clearer that it's just a callable

Comment (llllllllll) [inline codetransformer/transformers/literals.py]: what does it stand for?
